pkg/k8s/util.go

```diff
@@ -37,15 +38,80 @@ func GetHostnamesForService(svc service.MeshService, localNamespace bool) []stri
 	return hostnames
 }
 
-// GetServiceFromHostname returns the service name from its hostname
-func GetServiceFromHostname(host string) string {
-	// The service name is the first string in the host name for a service.
-	// Ex. service.namespace, service.namespace.cluster.local
-	service := strings.Split(host, ".")[0]
+// splitHostName takes a k8s FQDN (i.e. host) and retrieves the service name
+// as well as the subdomain (may be empty)
+func splitHostName(host string) (service string, subdomain string) {
+	serviceComponents := strings.Split(host, ".")
+
+	// The service name is usually the first string in the host name for a service.
+	// Ex. service.namespace, service.namespace.svc.cluster.local
+	// However, if there's a subdomain, we the service name is the second string.
+	// Ex. mysql-0.service.namespace, mysql-0.service.namespace.svc.cluster.local, mysql-0.service.namespace.svc.cluster.local
+	switch l := len(serviceComponents); {
+	case l == 1:
+		// e.g. service
+		service = serviceComponents[0]
+		subdomain = ""
+	case l == 2:
+		// e.g. service.namespace, mysql-0.service
+		// NOTE: There's no way to tell which component is the service or not, so we're going to keep the default behavior
+		// Thus, we have a requirement that users who want to reference statefulsets must include the namespace in the host name
```

Comment:
How would k8s resolve this? they likely have an order preference for search, and we could do something similar. Particularly by bringing these methods onto the individual clients/controllers, then we can correlate with the state of the world.

Reply:
CoreDNS is the standard implementation. I'd imagine they're deterministic. I'd avoid another dependency on that, but we could probably implement the same ordering.

The thing is, there's probably a negligible amount of users that have overlap between statefulsets where a conflict like this arises, so my thought is that allowing people to use the same semantics is a much better user experience. For the case where there is a conflict, we can try our best to have the same ordering as coreDNS, and for anybody else, we can add a caveat that they should expect our ordering.

I'd imagine this would leave us with near 0 users that attempt to use this feature and achieve an unexpected lookup.

Reply:
Another way of saying this, I'd imagine the percentage of users that try to use <pod>.<service_name> is much higher than the percentage of users that have a <pod>.<service_name> that conflicts with an existing <service_name>.<namespace>, so lets try to accommodate the former, and do our best for the latter

Reply:
Conflicts aren't really the issue here; remember, this function just takes a host name and says here's the subdomain and service name. This behavior has to remain consistent for all clusters, so there's no way to give best effort in the latter case. It's as if CoreDNS didn't ever support <service_name>.<namespace>; a DNS server has the benefit of just following NS records and returning an NXDOMAIN if it goes down the wrong path. In our case, if we said that `x.y` always meant `x = podName` and `y = service name`, then no OSM user would be able to use `service.namespace` and Envoy configs would break. 

The only way to do what you're suggesting is to either: 
1. do an actual DNS lookup or
2. add an additional heuristic to this function's parameters

Reply:
let me see if i can try to clarify what i'm saying with a small snippet

```
// l == 2
p1, p2 := strings.Cut(".")
// query for pods == p1, backed by service == p2.
validPodRef := k8sController.PodWithServiceExists(p1, p2)
// query for 
validServiceNSRef := k8sController.GetService(p1, p2)

switch {
case p1 && p2:
    // conflict resolution
case p1:
   ...
case p2:
   ....
}

```

does that make sense? I am essentially saying we should do a DNS lookup, but through the K8s API by querying for services/pods, not by adding a new dependency or performing an actual DNS query. I may be missing something else, so please let me  know!

Reply:
Ohhh I thought you were suggesting we could somehow do this with just strings; yeah, for this to work we'll need to pull in something else. I'll do that 👍🏾